internals/overlord/logstate/gatherer.go

```diff
@@ -0,0 +1,207 @@
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+	"gopkg.in/tomb.v2"
+)
+
+const parserSize = 4 * 1024
+
+// logGatherer is responsible for collecting service logs from a bunch of
+// services, and sending them to its logClient.
+// One logGatherer will run per log target. Its loop() method should be run
+// in its own goroutine. It can be stopped in a separate goroutine by calling
+// the stop() method.
+// The logGatherer will "flush" the client:
+//   - on a regular cadence (e.g. every 1 second)
+//   - when it is told to shut down.
+//
+// Its client may also flush itself when its internal buffer reaches a certain
+// size.
+type logGatherer struct {
+	targetName string
+	tickPeriod time.Duration
+	cancel     chan struct{}
+	client     logClient
+
+	entryCh     chan servicelog.Entry
+	pullers     map[string]*logPuller
+	pullersLock sync.Mutex
+}
+
+func newLogGatherer(target *plan.LogTarget) (*logGatherer, error) {
+	client, err := newLogClient(target)
+	if err != nil {
+		return nil, fmt.Errorf("could not create log client: %w", err)
+	}
+
+	return &logGatherer{
+		targetName: target.Name,
+		tickPeriod: 1 * time.Second,
+		cancel:     make(chan struct{}),
+		client:     client,
+		entryCh:    make(chan servicelog.Entry),
+		pullers:    map[string]*logPuller{},
+	}, nil
+}
+
+func (g *logGatherer) planChanged(pl *plan.Plan, buffers map[string]*servicelog.RingBuffer) {
+	g.pullersLock.Lock()
+	defer g.pullersLock.Unlock()
+
+	newPullers := map[string]*logPuller{}
+
+	for _, service := range pl.Services {
+		target := pl.LogTargets[g.targetName]
+		if !service.LogsTo(target) {
+			continue
+		}
+
+		if puller, ok := g.pullers[service.Name]; ok {
+			// Move over existing puller
+			newPullers[service.Name] = puller
+			delete(g.pullers, service.Name)
+			continue
+		}
+
+		buffer, bufferExists := buffers[service.Name]
+		if !bufferExists {
+			// We don't yet have a reference to the service's ring buffer
+			// Need to wait until serviceStarted
+			continue
+		}
+
+		// Create new puller
+		newPullers[service.Name] = newLogPuller(g.entryCh, buffer)
+	}
+
+	// Old pullers for now-removed services need to be shut down.
+	for svc, puller := range g.pullers {
+		err := puller.tomb.Killf("gatherer closed")
```

Comment:
gatherer didn't close, should be a different error.

Reply:
I'm not sure what you mean. We are killing the puller here because the gatherer has been asked to stop. Maybe I'm not properly understanding the semantics of the `Killf` argument?

Reply:
ah sorry, this is in `planChanged`. You're dead right